File: vuln-sandbox/csharp/cwe-918/ssrf.cs
// CWE-918: Server-Side Request Forgery — TRUE POSITIVE
// Rule: csharp/ssrf
// User-controlled host used in HttpClient.GetAsync URL.

using System.Net.Http;
using Microsoft.AspNetCore.Mvc;

public class ProxyController : Controller
{
    private readonly HttpClient _httpClient;

    public ProxyController(HttpClient httpClient)
    {
        _httpClient = httpClient;
    }

    [HttpGet("fetch")]
    public async Task<IActionResult> Fetch(string host)
    {
        // VULNERABLE: user-controlled host in URL
        var response = await _httpClient.GetAsync($"https://{host}/api/data");
        var content = await response.Content.ReadAsStringAsync();
        return Content(content);
    }
}

File: vuln-sandbox/rust/cwe-918/ssrf.rs
// VULNERABLE: rust-command-web-input — HTTP request to user-controlled URL
// Rule: rust-command-web-input | CWE-918 | Severity: CRITICAL

use reqwest::Client;

async fn fetch_url(user_url: &str) -> Result<String, reqwest::Error> {
    let client = Client::new();

    // VULNERABLE: user-controlled URL used for outbound HTTP request
    // An attacker can pass: http://169.254.169.254/latest/meta-data to access cloud metadata
    let response = client.get(user_url).send().await?;
    let body = response.text().await?;
    Ok(body)
}

#[tokio::main]
async fn main() {
    // Simulating user input from an HTTP request parameter
    let url = std::env::args().nth(1).unwrap_or_default();
    match fetch_url(&url).await {
        Ok(body) => println!("{}", body),
        Err(e) => eprintln!("error: {}", e),
    }
}
